test/integration/hds_integration_test.cc

```diff
@@ -875,5 +925,131 @@ TEST_P(HdsIntegrationTest, TestDefaultTimer) {
+
+  // Envoy sends a health check message to an endpoint
+  healthcheckEndpoints();
+
+  // Endpoint responds to the health check
+  host_stream_->encodeHeaders(Http::TestResponseHeaderMapImpl{{":status", "200"}}, false);
+  host_stream_->encodeData(1024, true);
+
+  // Receive updates until the one we expect arrives
+  waitForEndpointHealthResponse(envoy::config::core::v3::HEALTHY);
+
+  checkCounters(1, 2, 1, 0);
+
+  // Clean up connections
+  cleanupHostConnections();
+  cleanupHdsConnection();
+}
+
+// Health checks a single endpoint over TLS with HTTP/1
+TEST_P(HdsIntegrationTest, SingleEndpointHealthyTlsHttp1) {
+  // Change member variable to specify host streams to have tls transport socket.
+  tls_hosts_ = true;
+
+  initialize();
+
+  // Server <--> Envoy
+  waitForHdsStream();
+  ASSERT_TRUE(hds_stream_->waitForGrpcMessage(*dispatcher_, envoy_msg_));
+  EXPECT_EQ(envoy_msg_.health_check_request().capability().health_check_protocols(0),
+            envoy::service::health::v3::Capability::HTTP);
+
+  // Server asks for health checking
+  server_health_check_specifier_ =
+      makeHttpHealthCheckSpecifier(envoy::type::v3::CodecClientType::HTTP1, true);
+  hds_stream_->startGrpcStream();
+  hds_stream_->sendGrpcMessage(server_health_check_specifier_);
+  test_server_->waitForCounterGe("hds_delegate.requests", ++hds_requests_);
+
+  // Envoy sends a health check message to an endpoint
+  healthcheckEndpoints();
+
+  // Endpoint responds to the health check
+  host_stream_->encodeHeaders(Http::TestResponseHeaderMapImpl{{":status", "200"}}, false);
+  host_stream_->encodeData(1024, true);
+
+  // Receive updates until the one we expect arrives
+  waitForEndpointHealthResponse(envoy::config::core::v3::HEALTHY);
+
+  checkCounters(1, 2, 1, 0);
+
+  // Clean up connections
+  cleanupHostConnections();
+  cleanupHdsConnection();
+}
+
+// Attempts to health check a TLS endpoint over plaintext, which should fail.
+TEST_P(HdsIntegrationTest, SingleEndpointUnhealthyTlsMissingSocketMatch) {
+  // Make the endpoints expect communication over TLS.
+  tls_hosts_ = true;
+
+  initialize();
+
+  // Server <--> Envoy
+  waitForHdsStream();
+  ASSERT_TRUE(hds_stream_->waitForGrpcMessage(*dispatcher_, envoy_msg_));
+  EXPECT_EQ(envoy_msg_.health_check_request().capability().health_check_protocols(0),
+            envoy::service::health::v3::Capability::HTTP);
+
+  // Make the specifier not have the TLS socket matches, so it will try to connect over plaintext.
+  server_health_check_specifier_ =
+      makeHttpHealthCheckSpecifier(envoy::type::v3::CodecClientType::HTTP1, false);
+  server_health_check_specifier_.mutable_cluster_health_checks(0)
+      ->mutable_health_checks(0)
+      ->mutable_timeout()
+      ->set_seconds(0);
+  server_health_check_specifier_.mutable_cluster_health_checks(0)
+      ->mutable_health_checks(0)
+      ->mutable_timeout()
+      ->set_nanos(100000000); // 0.1 seconds
```

Comment:
Is there a possibility of racing here? We have a lot of issues with HDS integration test flakes, just want to check we're not adding another.

Reply:
@htuch, thanks for pointing this one out to me. I just pushed a commit that removes these sets, because they are unnecessary. They are leftover when I was originally thinking that it was going to be a TIMEOUT test. I removed these statements, and then ran following the [guide you provided](https://github.com/envoyproxy/envoy/tree/master/test/integration#reproducing-test-flakes), running 1000 times. It passed with no issues.

Just for verification, I also ran 1000x times while leaving these statements here, and they did not cause an issue. Better to remove them because they are not useful.